Fix compiler warnings

keeper opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling this template <3

If you have any questions related to MariaDB or you just want to
hang out and meet other community members, please join us on
https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed, what was it looking like before
   the change and how it's looking with this patch applied
3. Do you think this patch might introduce side-effects in
   other parts of the server?
-->
## Description
gcc-12.2.1 emits following warnings and are fixed:

`-Wdeprecated-declarations`:
- storage/connect/libdoc.cpp

`-Wmaybe-uninitialized`:
- storage/connect/tabext.cpp

This is based on logs building in a Fedora 37 container, and the CMake options being used was:
```
-DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DPLUGIN_TOKUDB=NO -DPLUGIN_MROONGA=NO -DPLUGIN_SPIDER=NO -DPLUGIN_OQGRAPH=NO -DPLUGIN_PERFSCHEMA=YES -DPLUGIN_SPHINX=NO
```
which also aligned with builder bot (except ccache).

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.

## How can this PR be tested?

Two tests were executed:
- build the source code and see if there's any warning
- run `./mtr --suite=main --force --parallel=auto` to find any regressions

and both passed.

This PR doesn't introduce any feature or behaviour changes so no new test is needed.

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
(Currently the earliest maintained branch is 10.3)
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2509) <br/>All committers have signed the CLA.

keeper commented:
> How did you determine that best_last_op= 0 was the right initialization value?

It was set to 0 before 0b47c126e31cddda1e94588799599e138400bcf8 so I assume the value is correct. I considered to set it to [`YYEOF`](https://github.com/MariaDB/server/blob/11.0/storage/innobase/include/pars0grm.h#L54) but I'm not sure if it's correct in context so I use the plain 0 just like before.

> `crc` is only used if `fake_event_header` doesn't return true (an error), and do_checksum is true. 

You are correct. I didn't dive into `fake_event_footer` and thought it will be used in the function uninitialized when `do_checksum` is false.

> Which compiler out of curiosity?

The warning comes from GCC7.

I'll remove the conflict and false positive fixes, thank you for the review!

keeper commented:
The issues in https://github.com/MariaDB/server/pull/2509#issuecomment-1447459513 has been fixed and the base branch is now `10.6`.

dr-m commented on storage/innobase/pars/pars0opt.cc:
```diff
@@ -553,7 +553,7 @@ opt_search_plan_for_table(
 	dict_index_t*	index;
 	ulint		n_fields;
-	ulint		best_last_op;
+	ulint		best_last_op= 0;
```
Can you please also add a comment that says which compiler would emit a warning? If I remember correctly, until fairly recently (starting somewhere between GCC 7 and 9), many `-Wmaybe-uninitialized` were being emitted for no reason.

keeper replied:
The warning is generated from gcc-7.5.0, and the comment is added accordingly.

dr-m reviewed (changes requested):
Please be more specific. The warning flag is applicable to very many versions of GCC and Clang. Which exact versions is this about? I am using GCC 12.2.0, and it does not warn anything. That is why I removed the redundant initialization in the first place.

keeper commented:
Hi, we've been revised the whole commit to target warnings emitted from gcc-12.2.0, and fix some warnings.

ottok commented:
I backported this to 10.11 and tested running the full MTR (including --big-test) on [amd64/arm64/ppc64el/s390x](https://launchpad.net/~mysql-ubuntu/+archive/ubuntu/debian-mariadb/+builds?build_text=&build_state=all), and I also included this in a package for which I ran extensive [install/upgrade testing on Debian](https://salsa.debian.org/mariadb-team/mariadb-server/-/pipelines/512904). I did not come across any issues introduced by this change, so all this testing is in favor of merging this.

grooverdan commented:
Checked 12.2.1 and came up with the same compile warnings as [fedora 37 in bb](https://buildbot.mariadb.org/#/builders/530/builds/4153) which covers none of these changes. Could it be they are all false positives?

ottok commented:
This was cherry-picked to the official Debian/Ubuntu package of MariaDB 10.11.2 in https://salsa.debian.org/mariadb-team/mariadb-server/-/commit/e1d59c1da8c396fe983c86a51848f0295418dfe9

ottok commented:
Can you @dr-m please re-review this?

keeper committed:
Fix compiler warnings

-Wdeprecated-declarations:
  Replace deprecated xmlXPathInit() with xmlInitParser()[1]

-Wmaybe-uninitialized:
  `fil1` was initialized to 0 in commit 0138220 to fix compiler
  warnings, but `fil2` was missed.

ref:
[1]: https://gnome.pages.gitlab.gnome.org/libxml2/devhelp/libxml2-xpath.html#xmlXPathInit

LinuxJedi commented:
> Can you @dr-m please re-review this?

Hi @ottok, three of us have now tried and cannot reproduce the original problem that this is supposed to fix. Which is making this difficult to review.

@keeper can you please give us more details on how you triggered these in the first place? I'm assuming you were using 10.6, but what OS and cmake options?

keeper commented:
> @keeper can you please give us more details on how you triggered these in the first place? I'm assuming you were using 10.6, but what OS and cmake options?

I was using gcc-12.2.0 with CMake flags:
```
-DCMAKE_CXX_STANDARD=11;-DWITH_INNODB_MEMCACHED=ON;-DPLUGIN_AUTH_GSSAPI=NO;-DPLUGIN_AUTH_GSSAPI_CLIENT=OFF;-DPLUGIN_AUTH_PAM=NO;-DPLUGIN_AUTH_PAM_V1=NO;-DPLUGIN_FEEDBACK=NO;-DPLUGIN_OQGRAPH=NO;-DPLUGIN_ROCKSDB=YES;-DPLUGIN_CASSANDRA=OFF;-DPLUGIN_AWS_KEY_MANAGEMENT=OFF;-DPLUGIN_COLUMNSTORE=NO;-DPLUGIN_MROONGA=NO;-DPLUGIN_S3=NO;-DPLUGIN_TOKUDB
=NO;-DWITH_EMBEDDED_SERVER=OFF;-DWITH_WSREP=OFF;-DWITH_LIBAIO=ON;-DWITH_INNODB_LZMA=ON;-DWITH_JEMALLOC=ON;-DWITH_NUMA=ON
```
under Amazon Linux 2, which may be the reason causing all these inconsistencies.

I'm now working on a revision base on gcc-12.2.1, building in Fedora 37 container, which align with MariaDB builder bot including compiler warnings (that I should've done earlier). I'm really sorry for all these confusions, and many thanks.

keeper commented:
I've force pushed my fork and rebased the changes to `10.11`. The new revision is aligned with builder bot environment and CMake options as close as possible.

The fixes in previous version are dropped completely, and two warnings (can also be observed in builder bot log) being addressed are `-Wdeprecated-declarations` and `-Wmaybe-uninitialize`.

illuusio commented:
Build these on GCC 13 and there is plenty room for warning squashing.  `storage/connect/tabext.cpp` has problem with `-Wmaybe-uninitialized` but there is plenty of `-Warray-bounds=` warnings also. Also this ` xmlXPathInit` warning is prensent.

Xpath init  deprication is from 2022 https://github.com/GNOME/libxml2/commit/670701075b969962db1b869507318e81e96910d1 but new parser initialization is present on Ubuntu 18.04 but is `xmlXPathInit` and `xmlInitParser` compatible as there is small amount time that I needed to use libxml2 and XPath.

keeper commented:
> Build these on GCC 13 and there is plenty room for warning squashing. `storage/connect/tabext.cpp` has problem with `-Wmaybe-uninitialized` but there is plenty of `-Warray-bounds=` warnings also. Also this ` xmlXPathInit` warning is prensent.
> 
I used gcc-13.1.1 in latest Fedora docker image and didn't see `-Wmaybe-uninitialized` in `tabext.cpp`, could you share what cmake option you used?

And for `Warray-bounds`, I believe it has been addressed in https://github.com/MariaDB/server/pull/2640 but still needs further discussions.

illuusio commented:
> I used gcc-13.1.1 in latest Fedora docker image and didn't see `-Wmaybe-uninitialized` in `tabext.cpp`, could you share what cmake option you used?
> 
> And for `Warray-bounds`, I believe it has been addressed in #2640 but still needs further discussions.

I've used Gcc 13.0.1 and your switches which were in other discussion
```
mkdir -p build
cd build
cmake -DCMAKE_CXX_STANDARD=11 -DWITH_INNODB_MEMCACHED=ON -DPLUGIN_AUTH_GSSAPI=NO -DPLUGIN_AUTH_GSSAPI_CLIENT=OFF -DPLUGIN_AUTH_PAM=NO -DPLUGIN_AUTH_PAM_V1=NO -DPLUGIN_FEEDBACK=NO -DPLUGIN_OQGRAPH=NO -DPLUGIN_ROCKSDB=YES -DPLUGIN_CASSANDRA=OFF -DPLUGIN_AWS_KEY_MANAGEMENT=OFF -DPLUGIN_COLUMNSTORE=NO -DPLUGIN_MROONGA=NO -DPLUGIN_S3=NO -DPLUGIN_TOKUDB=NO -DWITH_EMBEDDED_SERVER=OFF -DWITH_WSREP=OFF -DWITH_LIBAIO=ON -DWITH_INNODB_LZMA=ON -DWITH_JEMALLOC=ON -DWITH_NUMA=ON ..
```

keeper commented:
> I've used Gcc 13.0.1 and your switches which were in other discussion

I tested with gcc-13.0.1 (also in Fedora 38 container) and didn't see more `-Wmaybe-uninitialized` in `tabext.cc` other than that I fixed in the PR, nor in [buildbot/amd64-fedora-36 dry-run](https://buildbot.mariadb.org/#/builders/386/builds/5606).

Could you point out where I may be missing? Many thanks.

illuusio commented:
I've little bit tested this with different distros on docker: Fedora 36 GCC 12.x, openSUSE tumbleweed 13.1.x and Ubuntu 23.04.
 
Fedora 36 didn't find any with  GCC version 12.2.1 20221121.  openSUSE 13.1.1 20230522 and Ubuntu 23.04 GCC gcc version 12.2.0 (Ubuntu 12.2.0-14ubuntu2) find these:
```
Tiedostossa, joka sisällytettiin kohdasta /rainbow/country/server/include/my_sys.h:20,
                                 kohdasta /rainbow/country/server/include/waiting_threads.h:19,
                                 kohdasta /rainbow/country/server/sql/sql_class.h:25,
                                 kohdasta /rainbow/country/server/storage/connect/tabext.cpp:14:
In function ”int safe_strcpy(char*, size_t, const char*)”,
    inlined from ”virtual bool EXTDEF::DefineAM(PGLOBAL, LPCSTR, int)” at /rainbow/country/server/storage/connect/tabext.cpp:129:14:
/rainbow/country/server/include/m_string.h:258:52: varoitus: array subscript 4159 is outside array bounds of ”const char [56]” [-Warray-bounds=]
  258 |   if (dst[dst_size - 2] != '\0' && src[dst_size - 1] != '\0')
      |                                    ~~~~~~~~~~~~~~~~^
In function ”int safe_strcpy(char*, size_t, const char*)”,
    inlined from ”virtual bool TDBEXT::MakeSrcdef(PGLOBAL)” at /rainbow/country/server/storage/connect/tabext.cpp:375:15:
/rainbow/country/server/include/m_string.h:258:52: varoitus: array subscript 4159 is outside array bounds of ”const char [43]” [-Warray-bounds=]
  258 |   if (dst[dst_size - 2] != '\0' && src[dst_size - 1] != '\0')
      |                                    ~~~~~~~~~~~~~~~~^
In function ”int safe_strcpy(char*, size_t, const char*)”,
    inlined from ”virtual bool TDBEXT::MakeSrcdef(PGLOBAL)” at /rainbow/country/server/storage/connect/tabext.cpp:352:15:
/rainbow/country/server/include/m_string.h:258:52: varoitus: array subscript 4159 is outside array bounds of ”const char [43]” [-Warray-bounds=]
  258 |   if (dst[dst_size - 2] != '\0' && src[dst_size - 1] != '\0')
      |                                    ~~~~~~~~~~~~~~~~^
Tiedostossa, joka sisällytettiin kohdasta /usr/include/stdio.h:906,
                                 kohdasta /rainbow/country/server/include/my_global.h:289,
                                 kohdasta /rainbow/country/server/storage/connect/tabext.cpp:13:
In function ”int sprintf(char*, const char*, ...)”,
    inlined from ”virtual bool TDBEXT::MakeSrcdef(PGLOBAL)” at /rainbow/country/server/storage/connect/tabext.cpp:373:28:
/usr/include/bits/stdio2.h:30:34: varoitus: ”fil2” may be used uninitialized [-Wmaybe-uninitialized]
   30 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   31 |                                   __glibc_objsize (__s), __fmt,
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   32 |                                   __va_arg_pack ());
      |                                   ~~~~~~~~~~~~~~~~~
/rainbow/country/server/storage/connect/tabext.cpp: In member function ”virtual bool TDBEXT::MakeSrcdef(PGLOBAL)”:
/rainbow/country/server/storage/connect/tabext.cpp:333:34: huom: ”fil2” esiteltiin täällä
  333 |                 char *fil1 = 0, *fil2;
      |                                  ^~~~
In function ”int safe_strcpy(char*, size_t, const char*)”,
    inlined from ”int CONDFIL::Init(PGLOBAL, PHC)” at /rainbow/country/server/storage/connect/tabext.cpp:68:18:
/rainbow/country/server/include/m_string.h:258:52: varoitus: array subscript 4159 is outside array bounds of ”const char [19]” [-Warray-bounds=]
  258 |   if (dst[dst_size - 2] != '\0' && src[dst_size - 1] != '\0')
      |                                    ~~~~~~~~~~~~~~~~^
In function ”int safe_strcpy(char*, size_t, const char*)”,
    inlined from ”bool TDBEXT::_ZN6TDBEXT7MakeSQLEP7_globalb.part.0(PGLOBAL, bool)” at /rainbow/country/server/storage/connect/tabext.cpp:516:14:
/rainbow/country/server/include/m_string.h:258:52: varoitus: array subscript 4159 is outside array bounds of ”const char [23]” [-Warray-bounds=]
  258 |   if (dst[dst_size - 2] != '\0' && src[dst_size - 1] != '\0')
      |                                    ~~~~~~~~~~~~~~~~^
In function ”int safe_strcpy(char*, size_t, const char*)”,
    inlined from ”virtual bool TDBEXT::MakeCommand(PGLOBAL)” at /rainbow/country/server/storage/connect/tabext.cpp:605:15:
/rainbow/country/server/include/m_string.h:258:52: varoitus: array subscript 4159 is outside array bounds of ”const char [25]” [-Warray-bounds=]
  258 |   if (dst[dst_size - 2] != '\0' && src[dst_size - 1] != '\0')
      | 
```

keeper commented:
@illuusio 
For the warnings `Warray-bounds` caused by `include/m_string.h:258`, it was discussed in https://github.com/MariaDB/server/pull/2640, and so far there's no reliable way to improve it, except using `strlen()` which may impact performance.

And the warning `Wmaybe-uninitialized` at `storage/connect/tabext.cpp:333` has been fixed in this PR, please see https://github.com/MariaDB/server/pull/2509/commits/562690f537de0db551030a285301b47cd9fc55fc, many thanks!

LinuxJedi commented:
> @illuusio For the warnings `Warray-bounds` caused by `include/m_string.h:258`, it was discussed in #2640, and so far there's no reliable way to improve it, except using `strlen()` which may impact performance.

Correct, this could be classed as a bug in GCC. But turning off the warning masks real bugs. If we used `strlen()` it would only be used in situations where it would cause a performance penalty (long src strings). We haven't yet come up with a solution for this, but it is on my radar.

> And the warning `Wmaybe-uninitialized` at `storage/connect/tabext.cpp:333` has been fixed in this PR, please see [562690f](https://github.com/MariaDB/server/commit/562690f537de0db551030a285301b47cd9fc55fc), many thanks!

Many thanks for this.

LinuxJedi reviewed (approved):
Approved for inclusion in 11.2
